cardano-node/src/Cardano/CLI/Delegation/PoolRegistry.hs

```diff
@@ -0,0 +1,301 @@
+  then fail $ "Public key not all-lowercase: " <> s
+  else Right $ Id text
+  where s = unpack text
+
+duplicates :: (Ord a) => [a] -> Set a
+duplicates xs = snd $ foldl step (Set.empty, Set.empty) xs
+  where
+    step :: Ord a => (Set a, Set a) -> a -> (Set a, Set a)
+    step (seen, dups) x =
+      if Set.member x seen
+      then (seen, Set.insert x dups)
+      else (Set.insert x seen, dups)
+
+--------------------------------------------------------------------------------
+-- * Instances
+--
+instance FromJSON Submission where
+  parseJSON = AE.withObject "Submission" $ \v->
+    case validateFields v of
+      [] -> Submission
+              <$> v .: "id" <?> AE.Key "id"
+              <*> v .: "ticker" <?> AE.Key "ticker"
+              <*> v .: "cost" <?> AE.Key "cost"
+              <*> v .: "homepage" <?> AE.Key "homepate"
+              <*> v .: "margin" <?> AE.Key "margin"
+              <*> v .:? "pledge_address" <?> AE.Key "pledge_address"
+      xs -> fail $ List.unlines xs
+    where validateFields :: AE.Object -> [String]
+          validateFields v =
+            [ "Missing fields: " <> commaList missing
+            | not (Set.null missing) ] <>
+            [ "Unexpected fields: " <> commaList unexpected
+            | not (Set.null unexpected) ]
+            where
+              keys, mandatory, optional', missing, unexpected :: Set Text
+              keys = Set.fromList $ HMap.keys v
+              mandatory = Set.fromList ["id", "ticker", "cost", "homepage", "margin"]
+              optional' = Set.fromList ["pledge_address"]
+              missing = mandatory `Set.difference` keys
+              unexpected = keys `Set.difference` (mandatory <> optional')
+
+              commaList :: Set Text -> String
+              commaList = List.intercalate ", " . (show . unpack <$>) . Set.toList
+
+instance FromJSON Id where
+  parseJSON = AE.withText "Id" $
+    (either (fail . unpack) pure) . validatePublicKey
+
+instance FromJSON Ticker where
+  parseJSON = AE.withText "Ticker" $ \v ->
+    let s = unpack v
+    in if (length s >= 3 && length s <= 4
+           && all Char.isAsciiUpper s)
+    then pure $ Ticker v
+    else fail $ "Not an uppercase ASCII of length 3 or 4: " <> take 32 s
+
+instance FromJSON URI.URI where
+  parseJSON = AE.withText "Text" $ \v ->
+    case URI.parseAbsoluteURI (unpack v) of
+      Nothing -> fail $ "Not an absolute URI: " <> unpack v
+      Just x -> pure x
+
+instance FromJSON Cost where
+  parseJSON = _withBoundedScientific "Cost" $ \x ->
+    if not (x >= 0)
+    then fail "Value of cost must be: (x >= 0)"
+    else pure $ Cost x
+
+instance FromJSON Margin where
+  parseJSON = _withBoundedScientific "Margin" $ \x ->
+    if  not (x >= 0.0 && x <= 1.0)
+    then fail "Value of margin must be: (x >= 0.0 && x <= 1.0)"
+    else pure $ Margin x
+
+instance FromJSON PledgeAddress where
+  parseJSON = AE.withText "PledgeAddress" $ \v ->
+    case Bech32.decode v of
```

Comment:
We will need [decodeLenient](https://hackage.haskell.org/package/bech32-1.0.0/docs/Codec-Binary-Bech32.html#v:decodeLenient) here! Cardano addresses are typically longer than 90 characters :man_shrugging:

Reply:
Ack, will do!